### PR TITLE
Message unification part 1

### PR DESCRIPTION
**Pull request by madadam**
This PR contains some refactoring, cleanup and code reorganization in preparation for the message unification itself which will happen in a separate PR.

### CODE REVIEW

File: src/messages/mod.rs [hunk 1]
```diff
@@ -329,9 +329,9 @@ impl Debug for SignedRoutingMessage {
     }
 }
 
-impl Debug for MessageContent {
+impl Debug for RoutingVariant {
```
Comment (jeanphilippeD) [inline src/messages/mod.rs]: That should have moved with MessageContent to the new file in previous commit.

File: src/messages/direct.rs [hunk 2]
```diff
@@ -18,14 +18,15 @@ use crate::{
...
 use std::{
     fmt::{self, Debug, Formatter},
     hash::{Hash, Hasher},
-    mem,
+    ops::Deref,
 };
 
 /// Body of DirectVariant
-#[derive(Eq, PartialEq, Serialize, Deserialize)]
+#[derive(Eq, PartialEq, Hash, Serialize, Deserialize)]
```
Comment (jeanphilippeD) [inline src/messages/direct.rs]: not 100% sure that was a win,  remove fairly simple implementation, but traded off a less easy usage.

File: src/states/adult/mod.rs [hunk 3]
```diff
@@ -586,6 +585,15 @@ impl Base for Adult {
                 self.direct_msg_backlog.push((p2p_node, msg));
                 Ok(Transition::Stay)
             }
+            msg => {
```
Comment (jeanphilippeD) [inline src/states/adult/mod.rs]: This is not ideal, we probably want to keep the exhaustive list in these matches because we need to look at them everytime a variant is added or changed.

File: src/states/adult/mod.rs [hunk 4]
```diff
@@ -411,9 +410,9 @@ impl Adult {
         if self.in_location(msg.message_dst()) {
             let signed_msg = msg.take_or_deserialize_signed_routing_message()?;
             match &signed_msg.routing_message().content {
-                RoutingVariant::GenesisUpdate(info) => {
+                Variant::GenesisUpdate(info) => {
                     self.verify_signed_message(&signed_msg)?;
-                    return self.handle_genesis_update(info.clone());
+                    return self.handle_genesis_update((**info).clone());
                 }
                 _ => {
```
Comment (jeanphilippeD) [inline src/states/adult/mod.rs]: hence the reason to enumerate all, it is strange that for the direct message we choose to do something different (i.e not add to backlog).

File: src/states/elder/mod.rs [hunk 5]
```diff
@@ -1559,6 +1553,14 @@ impl Base for Elder {
                 );
                 self.direct_msg_backlog.push((p2p_node, msg));
             }
+            msg => {
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: Same

File: src/states/elder/mod.rs [hunk 6]
```diff
@@ -1559,6 +1553,14 @@ impl Base for Elder {
...
+                trace!(
```
Comment (jeanphilippeD) [inline src/states/elder/mod.rs]: The same decision was not made for `dispatch_routing_message` because match was not exhausitve, was it right?
(I guess it does not matter very much for this intermediate state because it cannot happen).

File: src/messages/mod.rs [hunk 7]
```diff
@@ -347,6 +351,11 @@ impl VerifyStatus {
     }
 }
 
+pub enum QueuedMessage {
```
Comment (jeanphilippeD) [inline src/messages/mod.rs]: Yes, that change seem to make quite a lot of sense to me.

File: src/messages/security_metadata.rs [hunk 8]
```diff
@@ -16,31 +16,7 @@ use crate::{
...
-use std::{
-    collections::BTreeSet,
-    fmt::{self, Debug, Formatter},
-};
-
-/// Metadata needed for verification of the sender.
-/// Contain shares of the section signature before combining into a BLS signature
-/// and into a FullSecurityMetadata.
-#[derive(Eq, PartialEq, Clone, Hash, Serialize, Deserialize)]
-pub struct PartialSecurityMetadata {
```
Comment (jeanphilippeD) [inline src/messages/security_metadata.rs]: It probably make more sense to fold that commit with the previous one since we are adding AccumulatingMessage to functionally replace PartialSecurityMetadata, and that removal change is pretty small.

File: src/location.rs [hunk 9]
```diff
@@ -6,77 +6,86 @@
...
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
+pub enum DstLocation {
+    /// Destination is a single node with the given name.
+    Node(XorName),
+    /// Destination are the nodes of the section whose prefix matches the given name.
+    Section(XorName),
+    /// Destination are the nodes whose name is matched by the given prefix.
+    Prefix(Prefix<XorName>),
+    /// Destination is the node at the `ConnectionInfo` the message is directly sent to.
+    Direct,
```
Comment (dirvine) [inline src/location.rs]: I think this should go as well. We will always have the connectioninfo to send the message to. So when it's a message to forward the receiver only needs to check dst == him or forward, but we cannot send a message to anyone we don't have the connection info for. So here we use Direct, but I think that itsef may be soemthing the `send` method gives to `quic-p2p` when it selects a target Feels like we perhaps mix this up a little here. Maybe the unification push I have atm but the actual wire format will need to include the `<SocketAddr>` of the target and we can ignore the Xorname there. So the routing message has no direct, but the `quic-p2p` message needs a SocketAddr and the message. Maybe that is the answer?

File: src/messages/security_metadata.rs [hunk 10]
```diff
@@ -18,18 +18,20 @@ use bincode::serialize;
 use serde::{Deserialize, Serialize};
 use std::fmt::{self, Debug, Formatter};
 
-/// Metadata needed for verification of the sender.
 #[derive(Eq, PartialEq, Clone, Hash, Serialize, Deserialize)]
-pub struct SectionSecurityMetadata {
-    pub proof: SectionProofSlice,
-    pub signature: bls::Signature,
+#[allow(clippy::large_enum_variant)]
+pub enum SecurityMetadata {
```
Comment (jeanphilippeD) [inline src/messages/security_metadata.rs]: Yes, flattening this enum make sense since it is simpler now.

File: src/messages/accumulating.rs [hunk 11]
```diff
@@ -104,14 +104,14 @@ impl AccumulatingMessage {
             }
         };
 
-        let security_metadata = SecurityMetadata::Section {
+        let src_authority = SrcAuthority::Section {
```
Comment (jeanphilippeD) [inline src/messages/accumulating.rs]: Yes, the rename seem pretty reasonable.

File: src/messages/accumulating.rs [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+            dst,
+            variant,
+            proof,
+            public_key_set,
+            signature_shares,
+        })
+    }
+
+    /// Add the signature shares of `other` into this message.
+    /// Note: currently no validation is performed that the messages have the same content.
```
Comment (dirvine) [inline src/messages/accumulating.rs]: Perhaps a big FIXME here to be addressed soon and we can look t verifying the content and sig at the same time and do away with a bad_sigs container.

File: src/location.rs [hunk 13]
```diff
@@ -6,77 +6,86 @@
...
-    Section(XorName),
-    /// A set of nodes with names sharing a common prefix - may span multiple `Section`s present in
-    /// the routing table or only a part of a `Section`
-    PrefixSection(Prefix<XorName>),
-    /// A single node
-    Node(XorName),
+/// Message source location.
+#[allow(clippy::large_enum_variant)]
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
+pub enum SrcLocation {
```
Comment (dirvine) [inline src/location.rs]: Is it worth having these as seperate files?

File: src/messages/accumulating.rs [hunk 14]
```diff
@@ -0,0 +1,299 @@
+// Copyright 2020 MaidSafe.net limited.
```
Comment (dirvine) [inline src/messages/accumulating.rs]: The file should be called accumulating_message perhaps?

File: src/messages/authority.rs [hunk 15]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 MaidSafe.net limited.
```
Comment (dirvine) [inline src/messages/authority.rs]: Again perhaps rename file src_authority ?

File: tests/mock_network/messages.rs [hunk 16]
```diff
@@ -90,14 +91,23 @@ fn send_and_receive() {
...
+        .filter(|n| n.inner.is_elder() && n.in_dst_location(&dst))
     {
         loop {
             match node.try_next_ev() {
-                Ok(Event::MessageReceived { content, src, dst }) => {
+                Ok(Event::MessageReceived { content, src, .. }) => {
                     request_received_count += 1;
                     if req_content == content {
-                        if let Err(err) = node.inner.send_message(dst, src, res_content.clone()) {
+                        let res_src = SrcLocation::PrefixSection(*node.our_prefix());
```
Comment (jeanphilippeD) [inline tests/mock_network/messages.rs]: This change look a bit out of place in this commit (Splitting Location into SrcLocation and DstLocation, why is it relevant  here. I'm also not clear what we are trying to do here.

File: src/location.rs (hunk 13 above)
Comment (dirvine) [inline src/location.rs]: I wonder (just question) is `Src` and `Dst` alone perhaps not enough for these names?

File: src/location.rs [hunk 17]
```diff
@@ -11,29 +11,18 @@ use crate::xor_space::{Prefix, XorName};
...
-    /// A single section whose prefix matches the given name
-    Section(XorName),
-    /// A set of nodes with names sharing a common prefix - may span multiple `Section`s present in
-    /// the routing table or only a part of a `Section`
-    PrefixSection(Prefix<XorName>),
-    /// A single node
+    /// A single node with the given name.
     Node(XorName),
+    /// A single section with the given prefix.
+    Section(Prefix<XorName>),
```
Comment (jeanphilippeD) [inline src/location.rs]: Not 100% sure since SrcLocation::Section means something diferent from DstLocation::Section, seem a bit strange.

File: src/location.rs (hunk 13 above)
Comment (jeanphilippeD) [reply src/location.rs]: SrcLocation is more explicit and does not really add noise when used in code. I'd prefer we keep the full thing.

File: src/location.rs (hunk 9 above)
Comment (jeanphilippeD) [reply src/location.rs]: Could be a follow up PR if that make sense, maybe?

File: src/messages/accumulating.rs [hunk 18]
```diff
@@ -31,18 +34,22 @@ pub struct AccumulatingMessage {
...
-        content: RoutingMessage,
+        src: Prefix<XorName>,
+        dst: DstLocation,
+        variant: Variant,
         section_share: &SectionKeyShare,
         public_key_set: bls::PublicKeySet,
         proof: SectionProofSlice,
     ) -> Result<Self> {
-        let bytes = serialize(&content)?;
+        let bytes = serialize(&(&src, &dst, &variant))?;
```
Comment (jeanphilippeD) [inline src/messages/accumulating.rs]: This split of one type into multiple does not seem quite right, especially as we serialize it like that with a specific order in multiple places. Would it make sense to still have a type, maybe one that serialize like it happens here?

File: src/messages/authority.rs [hunk 19]
```diff
@@ -6,37 +6,65 @@
...
 
 impl SrcAuthority {
+    pub fn location(&self) -> SrcLocation {
+        match self {
+            Self::Node { public_id, .. } => SrcLocation::Node(*public_id),
+            Self::Section { prefix, .. } => SrcLocation::Section(*prefix),
+        }
+    }
+
+    pub fn as_node(&self) -> Result<&PublicId> {
```
Comment (jeanphilippeD) [inline src/messages/authority.rs]: Not 100% sure about this interface and `as_section`, they seem easy to use incorrectly.

File: src/location.rs (hunk 13 above)
Comment (jeanphilippeD) [reply src/location.rs]: I did thought it made sense in one file when reviewing.

File: src/messages/direct.rs (hunk 2 above)
Comment (madadam) [reply src/messages/direct.rs]: The custom `Hash` impl becomes more annoying when we merge `DirectVariant` and `RoutingVariant` into one type. Anyway, ideally we would have `Hash` implemented for `parsec::{Request, Response}`, then none of this would be necessary.

File: src/chain/chain.rs [hunk 20]
```diff
@@ -1457,7 +1457,7 @@ impl Chain {
                 }
                 self.candidates(target_name)?
             }
-            Location::PrefixSection(ref prefix) => {
+            DstLocation::Prefix(ref prefix) => {
```
[the remark this answers is not in the document]
Comment (madadam) [reply src/chain/chain.rs]: I know you wanted to remove it but I would prefer to do it in a separate PR because it's more involved - required changes in the `NeighbourInfo` messages flow.

File: src/messages/direct.rs (hunk 2 above)
Comment (jeanphilippeD) [reply src/messages/direct.rs]: Make sense, it only require to add Hash to parsec NetworkEvent since we already implement it anyway in Routing.

File: src/location.rs (hunk 9 above)
Comment (madadam) [reply src/location.rs]: So `DstLocation::Direct` does several things:
1. It's for messages sent directly to a socket address we know, so no hops/relays take place
2. We don't know or don't care about the recipient's name
3. The destination is whoever actually receives the message (so `in_location(Direct)` is always `true`)

I added this variant to keep the "direct message" flows we currently have in order to keep the amount of changes in this PR somehow reasonable. I don't mind trying to addres this in a followup PR though.

File: src/messages/accumulating.rs (hunk 12 above)
Comment (madadam) [reply src/messages/accumulating.rs]: I'd say this is a low priority because in practice the signature accumulator calls this function only on messages that have identical hashes (and thus content). Might be still worth at least raising a github issue for it.

File: src/location.rs (hunk 13 above)
Comment (madadam) [reply src/location.rs]: Not sure, they are pretty small and related. Makes more sense to me to keep then in the same file.

File: src/messages/authority.rs (hunk 19 above)
Comment (madadam) [reply src/messages/authority.rs]: They simplify the message dispatch though. In what way could they be used incorrectly?

File: src/location.rs (hunk 13 above)
Comment (madadam) [reply src/location.rs]: Yeah, me too. It's bit more explicit without being that much more verbose.

File: tests/mock_network/messages.rs (hunk 16 above)
Comment (madadam) [reply tests/mock_network/messages.rs]: Previously we would just flip the src and dst when sending the response, but we can't do that anymore as they are different types. Also in later commit they actually have different variants so a simple mapping of src <-> dst won't work.

Comment (jeanphilippeD) [reply tests/mock_network/messages.rs]: Wouldn't it make sense to do it in the later commit then where it is necessary?
It just seem odd since we never use PrefixSection as a source so far (I think) (in later commit we start using the Prefix as the source, but I think we make it behave differently).

Comment (madadam) [reply tests/mock_network/messages.rs]: > Wouldn't it make sense to do it in the later commit then where it is necessary

But it is necessary here. I can't just take `SrcLocation` and pass it to where `DstLocation` is expected (which is what we did previously because they were the same types). Or am I misunderstanding what you mean?

File: src/location.rs (hunk 17 above)
Comment (madadam) [reply src/location.rs]: Should I rename it to `Prefix` perhaps?

File: tests/mock_network/messages.rs (hunk 16 above)
Comment (jeanphilippeD) [reply tests/mock_network/messages.rs]: I mean you can match the dst to a DstLocation::Section(name) => SrcLocation(name) no?

Comment (jeanphilippeD) [reply tests/mock_network/messages.rs]: Like you do for the res_dst.

File: src/location.rs (hunk 17 above)
Comment (jeanphilippeD) [reply src/location.rs]: Maybe, but we might loose the context that this is a Section?

Comment (jeanphilippeD) [reply src/location.rs]: Maybe the fact that they contain different type but named the same is less confusing. (like you did).

Comment (madadam) [reply src/location.rs]: So `PrefixSection` then? As it was?

Comment (jeanphilippeD) [reply src/location.rs]: It was meaning a different thing and is not particularly clear. I'd say lets forget that comment for now, I'm not sure the "fix" we have are helping.

File: src/chain/chain.rs [hunk 21]
```diff
@@ -910,7 +910,7 @@ impl Chain {
     /// `Location::Node` instead of the stored knowledge. Has no effect for other location types.
```
Comment (jeanphilippeD) [inline src/chain/chain.rs]: Minor: comment : Location -> DstLocation

File: src/chain/chain.rs [hunk 22]
```diff
@@ -1423,9 +1427,9 @@ impl Chain {
     ///     - if our name *is* the destination, returns an empty set; otherwise
     ///     - if the destination name is an entry in the routing table, returns it; otherwise
     ///     - returns the `N/3` closest members of the RT to the target
-    pub fn targets(&self, dst: &Location) -> Result<(Vec<&P2pNode>, usize), RoutingError> {
+    pub fn targets(&self, dst: &DstLocation) -> Result<(Vec<&P2pNode>, usize), RoutingError> {
```
Comment (jeanphilippeD) [inline src/chain/chain.rs]: Might be worth doing a search and replace for `Location::Section`, `Location::PrefixSection` ...

File: src/messages/accumulating_message.rs [hunk 23]
```diff
@@ -269,9 +269,13 @@ mod tests {
...
         RoutingMessage {
-            src: SrcLocation::Section(rng.gen()),
+            src: SrcLocation::Section(gen_prefix(rng)),
             dst: DstLocation::Section(rng.gen()),
             content: Variant::UserMessage(rng.sample_iter(Standard).take(6).collect()),
         }
     }
+
+    fn gen_prefix(rng: &mut MainRng) -> Prefix<XorName> {
+        Prefix::new(rng.gen_range(0, 4), rng.gen())
```
Comment (jeanphilippeD) [inline src/messages/accumulating_message.rs]: Any reason for the prefix to not be fully random?

File: src/messages/with_bytes.rs [hunk 24]
```diff
@@ -139,18 +139,21 @@ mod tests {
...
 
     #[test]
     fn serialise_and_partial_at_hop_message() {
         let mut rng = rng::new();
         let full_id = FullId::gen(&mut rng);
-        let msg = gen_message(&mut rng);
-
-        let signed_msg_org = unwrap!(SignedRoutingMessage::single_source(msg, &full_id,));
+        let msg = RoutingMessage {
+            src: SrcLocation::Node(*full_id.public_id().name()),
```
Comment (jeanphilippeD) [inline src/messages/with_bytes.rs]: Why the change from SrcLocation::Section to Node? Not sure it makes much difference but I wonder. (Particularly since all our hop messages have a Section as a source.

File: src/messages/accumulating_message.rs (hunk 23 above)
Comment (madadam) [reply src/messages/accumulating_message.rs]: Not really except to keep a potential log output short. Shouldn't affect anything in this test anyway.

File: src/messages/with_bytes.rs (hunk 24 above)
Comment (madadam) [reply src/messages/with_bytes.rs]: I don't believe it makes any difference for this test, but I though that since the message is `single_source` it would be more realistic.
